### ISSUE TITLE
Pubnub stops receiving messages after a seemingly random amount of hours

### ISSUE THREAD

**Issue by natekspencer** [ovirs/pyvivint#9]
Pubnub stops receiving messages after 8 hours, give or take, despite making consistent and repetitive calls to the vivint sky api. In this scenario, another client connecting doesn't seem to have any effect on the pubnub messages resuming and only seems to resume on restarting.

**Comment by ovirs** [comment]
For context - noticed the same issue in the past so definitely not related to the latest changes. For me is very inconsistent. Sometimes is working 3-4 days without an issue and sometimes stops receiving messages after just a few hours. Currently is still receiving messages after about 12h since I've updated.

I'm gonna try to decompile the android app over the weekend and see if I can get any inspiration from there, see how they handle the problem.

**Comment by natekspencer** [comment]
Yeah, I knew it wasn't a new problem. But figured I'd open an issue on it anyway so that we have a reference to it here.

It's weird that there is no "disconnect" or any other type of status update from pubnub on this. It just silently stops working. I also noticed that pubnub just released a new version (5.0.0) a few days ago. I've upgraded my local copy and will let that run to see if there is any difference.

**Comment by ovirs** [comment]
just wanted to follow up on this - I have not yet had time to take a look at the android app code :disappointed:. I will try to do it this coming weekend.

Just curios @natekspencer - what are you running HA on? I recently upgraded to a raspberry pi 4 and so far didn't have any issues with Pubnub. It's still receiving messages after more than a week (the longest I ever had it working). Not sure if it's pure luck lol or is related to the PI 4 upgrade.

Also, I don't have the cameras set up in hacs-vivint integration yet - they're on a different network than my HA instance for now. Not sure it that has anything to do with it either.

**Comment by natekspencer** [comment]
I'm also running mine on a Pi4. The first couple of days I had this running, it had those pubnub connection losses after 8 hours and then another one after like 12 hours. Since then, I've rebooted my HASS environment too many times to know whether it has been stable. I wonder if it does have something to do with my own network connection issues though.

I had the local code running for about 18 hours before I rebooted my router due to an update and it was fine for that time, but the code was killed because of loss of wifi. I then started my local code again today and my computer crashed, haha. So I haven't really been able to test longevity.

I started it again just now and am tracking average time between messages as well as max time between messages. I'm trying to determine if there is a "sweet spot" where we can "safely" say: if I haven't received a pubnub message in X amount of time, disconnect and then reconnect the pubnub service.

**Comment by balansse** [comment]
I had the same issue with the [HomeBridge Vivint plugin](https://github.com/balansse/homebridge-vivint) - I solved this by doing a call to refresh the Vivint system info (https://vivintsky.com/api/systems/${panelId}) every minute. This might be overkill but this solution works - I tested it for more than several months now.

**Comment by natekspencer** [comment]
I have it set up in the HACS plugin to refresh every 5 minutes to that same endpoint. It looks like mine is going on 3 days fine now, so maybe it was just a weird network hiccup that mine suffered. @balansse, do you ever try to resubscribe to the channel on occasion though on the off-chance that the subscription messages have stopped flowing?

**Comment by balansse** [comment]
I tried refreshing just pubnub subscription but that did not help - it
still stopped responding after some time. PubNub library for NodeJS does
the reconnection automatically after I subscribe to it - I’m just handling
the status updates and writing that to log if it happens.

**Comment by natekspencer** [comment]
I noticed, too, that you HAVE to hit the Vivint API endpoint or it just stops sending messages like you found where just resubscribing did nothing.

I was just wondering if you had a combination of the two. Like the "refresh systems every 1 minute" and a "if no messages received in ## minutes, attempt to resubscribe" or anything like that.

**Comment by balansse** [comment]
Well this sounds sophisticated - all I did was added a ping to refresh the system info and that solved the problem for me. But I agree that it is not the best approach to ping the Vivint API so often. 

During the times when Vivint Web Access was still working I tried to watch closely how it keeps the PubNub session alive and saw lots of small requests to PubNub and VivintSky API, one every minute. So maybe Vivint used the same solution from their end as well. Too bad Web Access is dead now.
